### PR TITLE
Parenthesize function expressions in expression position

### LINKED ISSUE #937: Function expressions in statement position need parentheses
E.g. the parentheses are necessary in

```js
(function(){} && 0);
```

I'll work on this one. Looks like there's existing infrastructure (`firstInStatement`); it's just not consistently applied. (In fact it's not applied under any conditions at the moment.) @vjeux (or whoever), thoughts on doing it similarly to #921 v.s. using the existing `firstInStatement`?

### PR DESCRIPTION
**Pull request by bakkot**
Fixes #937.

Also removes a lot of useless parentheses, especially around arrow functions.

### CODE REVIEW

File: src/fast-path.js [hunk 1]
```diff
@@ -246,9 +246,8 @@ FPp.needsParens = function(assumeExpressionContext) {
   }
 
   if (
-    ((parent.type === "ArrowFunctionExpression" && parent.body === node) ||
-    parent.type === "ExpressionStatement") &&
-    startsWithOpenCurlyBrace(node)
+    parent.type === "ArrowFunctionExpression" && parent.body === node && startsWithNoLookaheadToken(node, false)
```
Comment (vjeux) [inline src/fast-path.js]: nit: can you add an inline comment before boolean, otherwise it's confusing

```js
/* forbidFunctionAndClass */ true
```

Comment (bakkot) [inline src/fast-path.js]: Done.

**Comment by bakkot** [comment]
> I'm much more confident in this as it's written with an exhaustive way

If we wanted to be *really* exhaustive, we'd enumerate every node type and have the `default` case be an `unreachable()` assertion. But that's probably overkill.

File: tests/template/__snapshots__/jsfmt.spec.js.snap [hunk 2]
```diff
@@ -107,7 +107,7 @@ async function f() {
 b()\`\`;
 
 // \\"ClassExpression\\"
-(class {})\`\`;
+(class {}\`\`);
```
Comment (jlongster) [inline tests/template/__snapshots__/jsfmt.spec.js.snap]: I just happened to be skimming this PR :) This is a regression.

Comment (vjeux) [inline tests/template/__snapshots__/jsfmt.spec.js.snap]: Both are valid (well, they both throw at runtime, but parse the same), so it doesn't really matter which one we output. Since this is such an edge case, it's probably never going to be used in practice and whichever we output is fine.

**Comment by existentialism** [comment]
Nice work @bakkot!

File: tests/template/__snapshots__/jsfmt.spec.js.snap (hunk 2 above)
Comment (bakkot) [inline tests/template/__snapshots__/jsfmt.spec.js.snap]: Yeah, that was my thinking. It parses fine.
